File: src/main/java/leetcode/LC0386_Lexicographical_Numbers.java
package leetcode;

import java.util.ArrayList;
import java.util.List;

public class LC0386_Lexicographical_Numbers {
    public List<Integer> lexicalOrder(int n) {
        List<Integer> numbers = new ArrayList<>();
        for (int i = 1; i <= 9; i++) {
            generate(i, n, numbers);
        }
        return numbers;
    }

    private void generate(int current, int limit, List<Integer> result) {
        if (current > limit) {
            return;
        }

        result.add(current);

        for (int i = 0; i <= 9; i++) {
            int next = current * 10 + i;

            if (next <= limit) {
                generate(next, limit, result);
                continue;
            }

            break;
        }
    }
}

File: src/main/kotlin/LC0386LexicographicalNumbers.kt
class LC0386LexicographicalNumbers {
    fun lexicalOrder(n: Int): List<Int> {
        val result: MutableList<Int> = mutableListOf()
        for (i in 1..9) {
            addDigit(i, n, result)
        }
        return result
    }

    private fun addDigit(num: Int, n: Int, result: MutableList<Int>) {
        if (num <= n) {
            result.add(num)
            for (i in 0..9) {
                addDigit(Integer.valueOf("$num$i"), n, result)
            }
        }
    }
}